File: Algorithms/p236_Lowest_Common_Ancestor_of_a_Binary_Tree/p236_Lowest_Common_Ancestor_of_a_Binary_Tree_1.py
# Definition for a binary tree node.
# class TreeNode(object):
#     def __init__(self, x):
#         self.val = x
#         self.left = None
#         self.right = None

class Solution(object):
    def lowestCommonAncestor(self, root, p, q):
        """
        :type root: TreeNode
        :type p: TreeNode
        :type q: TreeNode
        :rtype: TreeNode
        """
        path = []
        found = [0]
        self.getNodePath(root,p,q,path,found)
        res = None
        for e in path:
            if e[1]==0:
                res = e[0]
            else:
                break
        return res
        
    def getNodePath(self, root, p, q, path, found):
        if root:
            path.append((root,found[0]))
            if root==p:
                found[0] += 1
            if root==q:
                found[0] += 1
            if found[0]<2:
                self.getNodePath(root.left,p,q,path,found)
            if found[0]<2:
                self.getNodePath(root.right,p,q,path,found)
            if found[0]<2:
                del path[-1]

File: Algorithms/p236_Lowest_Common_Ancestor_of_a_Binary_Tree/p236_Lowest_Common_Ancestor_of_a_Binary_Tree_1.cpp
/**
 * Definition for a binary tree node.
 * struct TreeNode {
 *     int val;
 *     TreeNode *left;
 *     TreeNode *right;
 *     TreeNode(int x) : val(x), left(NULL), right(NULL) {}
 * };
 */
class Solution {
public:
    TreeNode* lowestCommonAncestor(TreeNode* root, TreeNode* p, TreeNode* q) {
        vector<pair<TreeNode*,int>> path;
        int found = 0;
        getNodePath(root,p,q,path,found);
        TreeNode* res = 0;
        for(const auto& e : path){
            if(e.second==0){
                res = e.first;
            }else{
                break;
            }
        }
        return res;
    }
    void getNodePath(TreeNode* root, TreeNode* p, TreeNode* q, vector<pair<TreeNode*,int>>& path, int& found){
        if(root){
            path.push_back({root,found});
            if(root==p) ++found;
            if(root==q) ++found;
            if(found<2) getNodePath(root->left,p,q,path,found);
            if(found<2) getNodePath(root->right,p,q,path,found);
            if(found<2) path.pop_back();
        }
    }
};
